NickvisionTubeConverter.Shared/Models/Download.cs

```diff
@@ -123,91 +124,99 @@ public static async Task<string> GetVideoTitleAsync(string url)
-                }
-                try
-                {
-                    ytdlp.YoutubeDL(ytOpt).download(new string[1] {VideoUrl});
-                    IsDone = true;
-                    return true;
-                }
-                catch (Exception e)
+                using (Python.Runtime.Py.GIL())
                 {
-                    Console.WriteLine(e);
-                    IsDone = true;
-                    return false;
+                    dynamic ytdlp = Python.Runtime.Py.Import("yt_dlp");
+                    var hooks = new List<Action<Python.Runtime.PyDict>> { };
+                    hooks.Add(ProgressHook);
+                    var ytOpt = new Dictionary<string, dynamic> {
+                        {"quiet", true},
+                        {"ignoreerrors", "downloadonly"},
+                        {"merge_output_format", "mp4/webm/mp3/opus/flac/wav"},
+                        {"progress_hooks", hooks},
+                        {"postprocessor_hooks", hooks}
+                    };
+                    if (_fileType.GetIsAudio())
+                    {
+                        ytOpt.Add("format", "ba/b");
+                    }
+                    else if (_fileType == MediaFileType.MP4)
+                    {
+                        ytOpt.Add("format", "bv[ext=mp4]+ba[ext=m4a]/b[ext=mp4]");
+                    }
+                    else
+                    {
+                        ytOpt.Add("format", "bv+ba/b");
+                    }
+                    try
+                    {
+                        ytdlp.YoutubeDL(ytOpt).download(new string[1] { VideoUrl });
+                        IsDone = true;
+                        return true;
+                    }
+                    catch (Exception e)
+                    {
+                        Console.WriteLine(e);
+                        IsDone = true;
+                        return false;
+                    }
                 }
-            }
-        });
+            }, _cancellationToken.Token);
+        }
+        catch(TaskCanceledException)
+        {
+            IsDone = false;
+            return false;
+        }
     }
 
     /// <summary>
     /// Stops the download
     /// </summary>
     public void Stop() => _cancellationToken?.Cancel();
+
+    private void ProgressHook(Python.Runtime.PyDict entries)
+    {
+        using (Python.Runtime.Py.GIL())
+        {
+            var result = new DownloadProgressState()
+            {
+                Status = entries["status"].As<string>() switch
+                {
+                    "processing" => ProgressStatus.Processing,
+                    "downloading" => ProgressStatus.Downloading,
+                    _ => ProgressStatus.Other
+                },
+                Progress = entries.HasKey("downloaded_bytes") ? entries["downloaded_bytes"].As<double>() / entries["total_bytes"].As<double>() : 0,
+                Speed = entries.HasKey("speed") ? entries["speed"].As<double>() : 0
```

Comment:
if during download the speed is unknown (in case of connection problems for example), the key `speed` will exist, but `entries["speed"]` will be `None` (python's equivalent to null), that can't be converted to double, this will cause crash. Idk if `As<double?>()` will work, that's why I used try-catch.

Reply:
Works good, I tested it by manually setting `entries["speed"]` to `None`